### PR TITLE
SyntaxNodes: implement the hash function for syntax nodes.

### PR DESCRIPTION
**Pull request by nkcsgexi**
This hash function will concatenate all interesting pieces of information
of node definitions in a single string and call hash() on this string.

### CODE REVIEW

**Review by akyrtzi — CHANGES_REQUESTED** [review]
I have a recommendation to improve how the hash is calculated. Instead of concatenating strings you can use `hash`'s ability to hash a tuple and get the hash but "feeding" the values into the function, like this:

```
def hash_syntax_node(node):
	    # Hash into the syntax name and serialization code
	    result = hash((node.name, get_serialization_code(node.syntax_kind)))
	    for child in node.children:
	        # Hash into the expected child syntax
	        result = hash((result, child.syntax_kind))
	        # Hash into the child name
	        result = hash((result, child.name))
	        # Hash into whether the child is optional
	        result = hash((result, child.is_optional))
	    return result
```

**Comment by nkcsgexi** [comment]
@akyrtzi Done.

**Review by akyrtzi — APPROVED** [review]
👍

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

File: utils/gyb_syntax_support/__init__.py [hunk 1]
```diff
@@ -141,5 +143,28 @@ def dedented_lines(description):
     return textwrap.dedent(description).split('\n')
 
 
+def hash_syntax_node(node):
+    # Hash into the syntax name and serialization code
+    result = hash((node.name, str(get_serialization_code(node.syntax_kind))))
```
Comment (akyrtzi) [inline utils/gyb_syntax_support/__init__.py]: Oops sorry, missed a nitpick, you don't have to convert the integer to a string in order to hash it.
Also no need for `str(child.is_optional)` and `str(token.serialization_code)` below.

File: utils/gyb_syntax_support/__init__.py [hunk 2]
```diff
@@ -141,5 +143,28 @@ def dedented_lines(description):
...
+def hash_token_syntax(token):
+    # Hash into the token name and serialization code
+    return hash((token.name, str(token.serialization_code)))
+
+
 def calculate_node_hash():
-    return "abc"
+    result = 0
+    for node in SYNTAX_NODES:
+        result = hash((result, hash_syntax_node(node=node)))
```
Comment (akyrtzi) [inline utils/gyb_syntax_support/__init__.py]: Recommend `hash_syntax_node(node=node)` -> `hash_syntax_node(node)`, named arguments are not required in python, you usually put them for readability when there's a lot of them, but it's not offering any value here
Also remove named argument for `hash_token_syntax(token=token)` below.

**Comment by nkcsgexi** [comment]
Thanks! Addressed in https://github.com/apple/swift/pull/22469

**Pull request merged**